blockstore.go

```diff
@@ -3,121 +3,308 @@ package lmdbbs
 )
 
-func Open(path string) (*Blockstore, error) {
+type Options struct {
+	// Path is the directory where the LMDB blockstore resides. If it doesn't
+	// exist, it will be created.
+	Path string
+
+	// ReadOnly, if true, opens this blockstore in read-only mode.
+	ReadOnly bool
+
+	// NoSync disables flushing system buffers to disk immediately when
+	// committing transactions.
+	NoSync bool
+
+	// +++ DB sizing fields. +++ //
+	// InitialMmapSize is the initial mmap size passed to LMDB when
+	// creating/opening the environment.
+	InitialMmapSize int64
+
+	// MmapGrowthStepFactor determines the next map size when a write fails. The
+	// current size is multiplied by the factor, and rounded up to the next
+	// value divisible by os.Getpagesize(), to obtain the new map size, which is
+	// applied with mdb_env_set_mapsize.
+	MmapGrowthStepFactor float64
+
+	// MmapGrowthStepMax is the maximum step size by which we'll grow the mmap.
+	MmapGrowthStepMax int64
+	// --- DB sizing fields. --- //
+
+	// MaxReaders is the maximum amount of concurrent reader slots that exist
+	// in the lock table.
+	MaxReaders int
+}
+
+func Open(opts *Options) (*Blockstore, error) {
+	path := opts.Path
+	switch st, err := os.Stat(path); {
+	case os.IsNotExist(err):
+		if err := os.MkdirAll(path, 0777); err != nil {
+			return nil, fmt.Errorf("failed to create lmdb data directory at %s: %w", path, err)
+		}
+	case err != nil:
+		return nil, fmt.Errorf("failed to check if lmdb data dir exists: %w", err)
+	case !st.IsDir():
+		return nil, fmt.Errorf("lmdb path is not a directory %s", path)
+	}
+
+	pagesize := os.Getpagesize()
+	if pagesize < 0 {
+		pagesize = 4096 // being defensive and setting a safe value (4KiB).
+	}
+
+	// Validate mmap sizing parameters.
+	//
+	// InitialMmapSize cannot be negative nor zero, and must be rounded up to a multiple of the OS page size.
+	if v := opts.InitialMmapSize; v <= 0 {
+		log.Warnf("initial mmap size (%d) cannot be negative or zero; falling back to default: %d", v, DefaultInitialMmapSize)
+		opts.InitialMmapSize = DefaultInitialMmapSize
+	}
+	if v := roundup(opts.InitialMmapSize, int64(pagesize)); v != opts.InitialMmapSize {
+		log.Warnf("initial mmap size (%d) must be a multiple of the OS pagesize (%d); rounding up: %d", opts.InitialMmapSize, pagesize, v)
+		opts.InitialMmapSize = v
+	}
+
+	// MmapGrowthStepMax cannot be negative nor zero, and must be rounded up to a multiple of the OS page size.
+	if v := opts.MmapGrowthStepMax; v <= 0 {
+		log.Warnf("maximum mmap growth step (%d) cannot be negative or zero; falling back to default: %d", v, DefaultMmapGrowthStepMax)
+		opts.MmapGrowthStepMax = DefaultMmapGrowthStepMax
+	}
+	if v := roundup(opts.MmapGrowthStepMax, int64(pagesize)); v != opts.MmapGrowthStepMax {
+		log.Warnf("maximum mmap growth step (%d) must be a multiple of the OS pagesize (%d); rounding up: %d", opts.MmapGrowthStepMax, pagesize, v)
+		opts.MmapGrowthStepMax = v
+	}
+
+	if v := opts.MmapGrowthStepFactor; v <= 1 {
```

Comment:
and here, we need something to mean default (0 for unset field?)

Reply:
0 is unset/default, as is 1 or anything below. I can remove the warns, though. This absolutely must be greater than 1 because otherwise the mmap will never grow.

Reply:
fair enough.